# Deriving New Variable

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

In [2]:
#defining a path
path=r'C:\Users\banda\Instacart Basket Analysis'

In [3]:
#importing pickle file
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data','Prepared data','ords_prods_merge.pkl'))

In [9]:
#creat new subset that only includes forst 1 million of the dataframe
df = ords_prods_merge[:1000000]

In [5]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge  
0         Soda        77              7     9.0   both  
1         Soda        77              7     9.0   both  
2         Soda        77              7     9.0   both  
3         Soda        77              7     9.0   both  
4         Soda        77              7     9.0   both

In [6]:
ords_prods_merge.shape

(32404859, 14)

In [10]:
#defining function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [11]:
#where the lable be returned : apply the function
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\banda\AppData\Local\Temp\ipykernel_3664\3677513468.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [12]:
df.shape

(1000000, 15)

In [13]:
df.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge        price_range  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [14]:
#checking the frequency of Prace_range column
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [15]:
#looking at the mas value in price column
df['prices'].max()

14.8

Using the loc () funtion with df dataframe

In [16]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\banda\AppData\Local\Temp\ipykernel_3664\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [17]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [18]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [19]:
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

using Loc() funtion for the entire dataframe, runs faster, can be use in large dataframe

In [22]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [24]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [23]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [25]:
#frequency check (When dropna is set to False, it includes the count of missing values (NaN) in the result. )
ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

In [26]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  
0         Soda        77              7     9.0   both  Mid-range product  
1         Soda        77              7     9.0   both  Mid-range product  
2         Soda        77              7     9.0   both  Mid-range product  
3         Soda        77              7     9.0   both  Mid-range product  
4         Soda        77              7     9.0   both  Mid-range product

In [28]:
#fequency check on orders data of the week. (Printing the frequency of a column will quickly inform you which values appear most often within that column)
ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [57]:
#for_loop function for creating a busiest day 
result = [] #his will act as the empty shell, into which you can place the results from your loop

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [58]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [59]:
#create a new column within your ords_prods_merge dataframe and set it equal to result
ords_prods_merge['busiest_day']=result

In [60]:
#frequency check
ords_prods_merge['busiest_day'].value_counts(dropna=False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

#task
they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [61]:
# creating a new column using loop funtion
X_result = []

for value in ords_prods_merge["orders_day_of_week"]:
    if value == 0 or value == 1:
        X_result.append("Busiest days")
    elif value == 4 or value == 3:
        X_result.append("Slowest days")
    else:
        X_result.append("Regularly days")

In [62]:
X_result

['Regularly days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Regularly days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Busiest days',
 'Regularly days',
 'Regularly days',
 'Busiest days',
 'Busiest days',
 'Regularly days',
 'Regularly days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Regularly days',
 'Regularly days',
 'Busiest days',
 'Regularly days',
 'Regularly days',
 'Busiest days',
 'Regularly days',
 'Regularly days',
 'Regularly days',
 'Slowest days',
 'Regularly days',
 'Slowest days',
 'Busiest days',
 'Busiest days',
 'Regularly days',
 'Slowest days',
 'Slowest days',
 'Regularly days',
 'Regularly days',
 'Busiest days',
 'Busiest days',
 'Regularly days',
 'Busiest days',
 'Busiest days',
 'Slowest days',
 'Regularly days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest days',
 'Busiest 

In [75]:
#create a new column within your ords_prods_merge dataframe and set it equal to X_result
ords_prods_merge['busiest_days']=X_result

In [76]:
#checking the fequency of the result
ords_prods_merge['busiest_days'].value_counts(dropna=False)

busiest_days
Regularly days    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

Check the values of this new column for accuracy. Note any observations in markdown format.
result shows as such busiest_day
Regularly days    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64 busiest days contains saturday and sunday and slowest days contains tuesdays and wednesday

When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [68]:
ords_prods_merge['order_hour_of_day'].value_counts(dropna=False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

In [70]:
# creating a new column with most orders, average orders and fewest orders using loop funtion
Y_result=[]
for value in ords_prods_merge['order_hour_of_day']:
    if value == 10:
        Y_result.append("Most orders")
    elif value == 3:
        Y_result.append("Fewest orders")
    else:
        Y_result.append("Average orders")

In [71]:
#create a new column within your ords_prods_merge dataframe and set it equal to Y_result
ords_prods_merge['busiest_period_of_day']=Y_result

In [72]:
Y_result

['Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Most orders',
 'Average orders',
 'Average orders',
 'Average orders',
 'Average 

In [73]:
#checking frequency
ords_prods_merge['busiest_period_of_day'].value_counts(dropna=False)

busiest_period_of_day
Average orders    29591818
Most orders        2761760
Fewest orders        51281
Name: count, dtype: int64

comparing the results frquencies
 order_hour_of_day                                     
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64
busiest_period_of_day
Average orders    29591818
Most orders        2761760
Fewest orders        51281
Name: count, dtype: int64


In [77]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      busiest_day      least_busy busiest_period_of_day    busiest_days  
0  Regularly days  Regularly days        Average orders  Regularly days  
1    Slowest days    Slowest days        Average orders    Slowest days  
2    Slowest days    Slowest days        Average orders    Slowest days  
3    Slowest days    Slowest days        Average orders    Slowest days  
4    Slowest days    Slowest days        Average orders    Slowest days

In [78]:
ords_prods_merge = ords_prods_merge.drop(columns = ['busiest_day','least_busy'])

In [79]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

  busiest_period_of_day    busiest_days  
0        Average orders  Regularly days  
1        Average orders    Slowest days  
2        Average orders    Slowest days  
3        Average orders    Slowest days  
4        Average orders    Slowest days

In [80]:
ords_prods_merge.shape

(32404859, 17)

In [81]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data','ords_prods_merge_1.pkl'))